### ISSUE TITLE
Updating `npm` moves library directory from `n`

### ISSUE THREAD

**Issue by SamuelMarks** [tj/n#294]
```
$ n stable
[…]
$ which npm
/usr/local/n/versions/node/0.12.7/bin/npm
$ npm list -g
/usr/local/n/versions/node/0.12.7/lib
└─┬ npm@2.11.3
  ├── abbrev@1.0.7
[…]
$ npm -g upgrade
/usr/local/n/versions/node/0.12.7/bin/npm -> /usr/local/n/versions/node/0.12.7/lib/node_modules/npm/bin/npm-cli.js
npm@2.13.1 /usr/local/n/versions/node/0.12.7/lib/node_modules/npm
$ npm list -g
/usr/local/lib
└── (empty)
```

**Comment by maxrimue** [comment]
Have you changed any environment variables? Node.js/npm should normally not run in `/usr/local/n/versions/...` but in `/usr/local/`, and n should install Node.js/npm there.
Running `n latest` + `which npm` returns `/usr/local/bin/npm` for me.

**Comment by SamuelMarks** [comment]
@maxrimue: So I haven't installed `npm` or `node` on my OS. Then I run:
```
$ git clone https://github.com/visionmedia/n
[…]
$ cd n
$ make install
[…]
$ n stable
[…]
```
Then I update my `PATH` variable to include:
```
/usr/local/n/versions/node/0.12.7/bin
```

How am I meant to add `node` and `npm` to my `PATH`?

**Comment by maxrimue** [comment]
You don't need to add node and npm to your path. By default, n will install Node.js into /usr/local/, which will automatically make node and npm accessible globally on your system.

**Comment by SamuelMarks** [comment]
So I removed the `PATH` updating, and here's the output:
```
$ which node
/usr/local/bin/node
$ which n
/usr/local/bin/n
$ which npm
$ echo $?
1
```

TL;DR: `npm` wasn't copied/linked to `/usr/local/bin/npm`.

**Comment by maxrimue** [comment]
Did you install Node.js with n without modifying any path variable?

**Comment by SamuelMarks** [comment]
Yes

**Comment by maxrimue** [comment]
If you `export npm=/usr/local/bin/npm` and then call, for example, `n 0.12.2`, does the path of npm stay the same, `/usr/local/bin/npm`, or does it change again?

**Comment by SamuelMarks** [comment]
Doesn't look like the `npm` binary is copied or linked across:

```
$ env -i bash
$ n stable

     install : node-v0.12.7
       mkdir : /usr/local/n/versions/node/0.12.7
       fetch : https://nodejs.org/dist/v0.12.7/node-v0.12.7-linux-x64.tar.gz
   installed : v0.12.7

$ export npm=/usr/local/bin/npm
$ which npm
$ echo $?
1
$ stat /usr/local/bin/npm
stat: cannot stat '/usr/local/bin/npm': No such file or directory
```

**Comment by maxrimue** [comment]
@SamuelMarks But the Node binary is? To be honest, I don't know what's going on there

**Comment by tedgaydos** [comment]
Try running this to fix your npm install.

`curl -L https://npmjs.com/install.sh | sh`

Why did you need to upgrade? Usually the version that comes packaged with node is enough. The way `n` works, it will always set the `npm` version to the copy stored in `$N_PREFIX/n/versions`. So if you upgrade and then switch the active version, your upgrade will get overwritten. If you really want to upgrade, you will need to to a local upgrade in the specific version folder and then reactivate it.

**Comment by SamuelMarks** [comment]
@tedgaydos TBH I don't want to run that command. I always want to be able to grab the latest versions of `npm` and `node` (and run `npm` without `sudo`). In order to do this, I was thinking something like `n` would facilitate.

@maxrimue Correct, the `node` binary is being copied across properly, but not the `npm` binary.

**Comment by maxrimue** [comment]
@SamuelMarks There we go, that's the problem, you installed npm manually. npm nowadays comes bundled with Node.js and shouldn't be installed manually, that's because certain npm versions aren't compatible with certain node versions. Also, I don't see what you are trying to achieve with this, using the latest node will you give you a version of npm that is new enough, there aren't any security problems nor incredible new features. My advise would be to remove that npm, and try to install node with n again, then you should have the bundled npm with each node version. If you're not willing to do this, you have to accept that bad things could happen when using an older node that might not be compatible with the manually installed version of npm you have.

**Comment by SamuelMarks** [comment]
Been using one of the answers from this question for ages:
- [**Install Node.js to install n to install Node.js?**](http://stackoverflow.com/q/19451851)

**Comment by Jonnymcc** [comment]
This PR may fix the problem experienced here https://github.com/tj/n/pull/364

**Comment by shadowspawn** [comment]
This issue has not had any activity in over six months. It isn't likely to get acted on due to this report.

Feel free to open a new issue if it comes up again, with new information and renewed interest.
 
Thank you  for your contributions.
